rcl_action/test/rcl_action/test_action_server.cpp

```diff
@@ -0,0 +1,738 @@
+        goal_info.uuid[j] = static_cast<uint8_t>(i + j);
+      }
+      goal_info.stamp.sec = i;
+      goal_handle = rcl_action_accept_new_goal(&this->action_server, &goal_info);
+      ASSERT_NE(goal_handle, nullptr) << rcl_get_error_string().str;
+    }
+  }
+
+  void TearDown()
+  {
+    TestActionServer::TearDown();
+  }
+
+  const int num_goals = 42;
+};
+
+TEST_F(TestActionServerCancelPolicy, test_action_process_cancel_request_all_goals)
+{
+  // Request to cancel all goals
+  rcl_action_cancel_request_t cancel_request = rcl_action_get_zero_initialized_cancel_request();
+  rcl_action_cancel_response_t cancel_response = rcl_action_get_zero_initialized_cancel_response();
+  rcl_ret_t ret = rcl_action_process_cancel_request(
+    &this->action_server, &cancel_request, &cancel_response);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  EXPECT_NE(cancel_response.msg.goals_canceling.data, nullptr);
+  ASSERT_EQ(cancel_response.msg.goals_canceling.size, (size_t)this->num_goals);
+  rcl_action_goal_info_t * goal_info_out;
+  for (int i = 0; i < this->num_goals; ++i) {
+    goal_info_out = &cancel_response.msg.goals_canceling.data[i];
+    for (int j = 0; j < 16; ++j) {
+      EXPECT_EQ(goal_info_out->uuid[j], static_cast<uint8_t>(i + j));
+    }
+  }
+}
+
+TEST_F(TestActionServerCancelPolicy, test_action_process_cancel_request_single_goal)
+{
+  // Request to cancel a specific goal
+  rcl_action_cancel_request_t cancel_request = rcl_action_get_zero_initialized_cancel_request();
+  for (int i = 0; i < 16; ++i) {
+    cancel_request.goal_info.uuid[i] = static_cast<uint8_t>(i + 2);
+  }
+  rcl_action_cancel_response_t cancel_response = rcl_action_get_zero_initialized_cancel_response();
+  rcl_ret_t ret = rcl_action_process_cancel_request(
+    &this->action_server, &cancel_request, &cancel_response);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  EXPECT_NE(cancel_response.msg.goals_canceling.data, nullptr);
+  ASSERT_EQ(cancel_response.msg.goals_canceling.size, 1u);
+  rcl_action_goal_info_t * goal_info = &cancel_response.msg.goals_canceling.data[0];
+  for (int i = 0; i < 16; ++i) {
+    EXPECT_EQ(goal_info->uuid[i], static_cast<uint8_t>(i + 2));
+  }
+}
+
+TEST_F(TestActionServerCancelPolicy, test_action_process_cancel_request_by_time)
+{
+  // Request to cancel all goals at and before a specific time
+  rcl_action_cancel_request_t cancel_request = rcl_action_get_zero_initialized_cancel_request();
+  cancel_request.goal_info.stamp.sec = 21;
+  rcl_action_cancel_response_t cancel_response = rcl_action_get_zero_initialized_cancel_response();
+  rcl_ret_t ret = rcl_action_process_cancel_request(
+    &this->action_server, &cancel_request, &cancel_response);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  EXPECT_NE(cancel_response.msg.goals_canceling.data, nullptr);
+  ASSERT_EQ(cancel_response.msg.goals_canceling.size, 22u);
+  rcl_action_goal_info_t * goal_info_out;
+  for (int i = 0; i < 22; ++i) {
```

Comment:
@jacobperron here and below, consider using `cancel_response.msg.goals_canceling.size` instead of a literal.

Reply:
Done 69c0287